### ISSUE TITLE
Running `nerdctl build` with Rancher Desktop fails to pull base images from custom registry with self-signed cert

### ISSUE THREAD

**Issue by stanleymho** [rancher-sandbox/rancher-desktop#909]
**For bugs, describe what you're seeing**

Using `nerdctl build` with Rancher Desktop on macOS involves pulling a base image from a custom registry which uses self-signed corporate cert, and the error is x509: certificate signed by unknown authority. I have the root CA certs in KeyChain as well as under /usr/local/share/ca-certificates on my host machine. I understand that Rancher Desktop has recently added support for installing the host CA certs into k3s under the cover. However, when I check the BuildKit instance running in the kube-image namespace in k3s under the cover of Rancher Desktop, it doesn’t seem to have the corporate root CA certs imported from the host machine.

**To Reproduce**
Steps to reproduce the behaviour:

```
$ nerdctl build -f Dockerfile .
```

**Result**
```
[+] Building 0.4s (3/3) FINISHED                                                                                                                                                            
 => [internal] load build definition from Dockerfile                                                                                                                            0.1s
 => => transferring dockerfile: 38B                                                                                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                        0.0s
 => ERROR [internal] load metadata for foobar.com/myimage:tag                                                                                                             0.2s
------
 > [internal] load metadata for foobar.com/myimage:tag
------
error: failed to solve: failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to do request: Head https://foobar.com/v2/myimage/manifests/tag: x509: certificate signed by unknown authority
FATA[0000] unrecognized image format            
```

**Setup (please complete the following information):**
 - OS: macOS 11.6.1
 - Rancher Desktop version: 0.6.1
 - Kubernetes version: 1.21.5

**Comment by jandubois** [comment]
I believe this is due to `kim` running `buildkitd` inside a container, so it doesn't see the certs installed on the guest.

I'm not aware of a way to provide `kim` with additional certs, so this probably cannot be fixed until we run `buildkitd` natively without `kim` (so will continue to be a problem in 0.7.0).

I expect this to work automatically with `dockerd` in 0.7.0, but of course needs verification.

**Comment by dweomer** [comment]
My idea is to add some configurability to `kim` to take tls secrets and drop them under `/etc/docker/certs.d` in the `buildkitd` container: https://docs.docker.com/engine/security/certificates/

**Comment by hypnoce** [comment]
On pure lima, this is how I fixed it : 
```
# Make 'myreg' an insecure registry
lima sh  << EOF
cat << EOM > ~/.config/buildkit/buildkitd.toml
[registry."myreg.mydomain.com"]
  insecure = true
EOM
EOF
```
it did not work on rancher desktop.

**Comment by shubb30** [comment]
I found a workaround, by injecting the corp certs into a secret, and then mounting that into the builder DaemonSet.

```
$ kubectl create secret generic -n kube-image corp-certs \
                             --from-file=root_ca="/Users/me/certificates/Root-CA.pem" \
                             --from-file=subCA="/Users/me/certificates/subCA.pem"
secret/corp-certs created
```

Then edit the daemonset: `kubectl edit -n kube-image daemonset.apps/builder`

Add the secret as a volume:

```
      volumes:
      ...
      - name: corp-certs
        secret:
          defaultMode: 420
          secretName: corp-certs
```

Add a volumemount:

```
        volumeMounts:
        ...
        - mountPath: /usr/local/share/ca-certificates/
          name: corp-certs
          readOnly: true
```

Add a lifecycle postStart to update the CA certs:

```
        image: docker.io/moby/buildkit:v0.8.3
        imagePullPolicy: IfNotPresent
        lifecycle:
          postStart:
            exec:
              command:
              - update-ca-certificates
```

And now, nerdctl/kim can build an image using an image in my private image repo.

**Comment by jandubois** [comment]
> I found a workaround, by injecting the corp certs into a secret, and then mounting that into the builder DaemonSet.

@shubb30 Cool, thanks for sharing!

**Comment by gaktive** [comment]
This should be in the release notes once we get the docs in some shape, along with the workaround.

@btat would you like a `release-note` label on things like this?

**Comment by btat** [comment]
> This should be in the release notes once we get the docs in some shape, along with the workaround.
> 
> @btat would you like a `release-note` label on things like this?

Yes, that'd be good.

**Comment by zimbres** [comment]
> I found a workaround, by injecting the corp certs into a secret, and then mounting that into the builder DaemonSet.

I'm using version 0.7.1, is it possible that something was changed? Even if loading a CA in the builder container, I cannot connect to a local registry.
I had checked and the cert is properly loaded into ca-certificates.crt

**Comment by shubb30** [comment]
I upgraded my RD to 0.7.1, and saw that the Deployment got updated and removed the changes.  I had to edit the deployment and add them back, but after doing that it is still working for me.

Are you able to `curl` to your registry from the `builder` pod?  You'll need to add it with `apk add curl`.  Mine is able to reach my registry without errors.  Make sure you have all of the root and intermediate CAs included in the secret.

If you are going to release note this as a workaround, you'll need to add that the Deployment will need to be updated whenever you upgrade Rancher Desktop.  This makes it less user-friendly.

**Comment by zimbres** [comment]
Hi, just tried with success!

Just after a wrote my message, I change for dockerd runtime, in this case the "etc/docker/daemon.json" aproach for insecure registry works fine.

Now to test by curl like you said I revert back for containerd and with the same steps, worked. Not sure why.

Thank you!

**Comment by drenze-athene** [comment]
> I found a workaround, by injecting the corp certs into a secret, and then mounting that into the builder DaemonSet.

So...when I restart `rancher-desktop`, it looks as if my updated daemonset gets overwritten?

**Comment by shubb30** [comment]
You are correct, the workaround only works until Rancher Desktop is restarted.  
@jandubois, @dweomer, I hope Rancher has a plan to fix this before January 31, when Docker Desktop can no longer be used.  Rancher Desktop is the best replacement I have found, but this issue makes it unusable for developers.

**Comment by drenze-athene** [comment]
> You are correct, the workaround only works until Rancher Desktop is restarted.

Thanks. I wanted to make certain I wasn't missing something.

> @jandubois, @dweomer, I hope Rancher has a plan to fix this before January 31, when Docker Desktop can no longer be used. Rancher Desktop is the best replacement I have found, but this issue makes it unusable for developers.

Seems to me that the best way to do this would be to incorporate these updates into the `builder` daemonset specification, along with a secret that does NOT get overwritten on reload.

In the meantime, as a workaround for the workaround, I've saved a copy of the modified daemonset specification locally and apply it when I need to do builds. Not optimal, but it's fairly quick to apply.

**Comment by stanleymho** [comment]
Is it feasible for Rancher Desktop or kim to start the builder daemonset by recognizing the root CA certificate on the host machine automatically?

I understand that manual workaround is possible. That said, for any company with sizable number of developers, asking each developer to apply the workaround manually and reinstate the workaround after each restart is not a feasible and scalable solution.

**Comment by jandubois** [comment]
> Is it feasible for Rancher Desktop or kim to start the builder daemonset by recognizing the root CA certificate on the host machine automatically?

Once #1182 is merged, `buildkitd` will be run as a regular service, and communication will be via a unix socket, which doesn't require a certificate (iiuc).

There should be a 1.0.0-beta release by early next week; we would appreciate if you could confirm that the problem has been fixed by that change.

**Comment by crumdev** [comment]
I'm still unable to pull images from a company registry with a Custom CA. I'm on RD version 1.0.0

**Comment by wstr-ncrum** [comment]
Ok so I switched from the containerd runtime to dockerd in the Rancher Desktop preferences, connected to the rancher-desktop WSL instance on my windows machine and created the `/etc/docker/daemon.json` file with an entry saying my Registry is insecure like in the link below and now I'm able to pull images from our registry.

Since this isn't ideal though I'm going to try and copy our root ca into the WSL instance and trust it next. Just wanted to provide an interim update.

https://stackoverflow.com/questions/42211380/add-insecure-registry-to-docker

**Comment by wstr-ncrum** [comment]
Ok still using dockerd I removed the `\\wsl$\rancher-desktop\etc\docker\daemon.json\` quit and restart rancher-desktop and confirmed that I was unable to pull images again. Then I copied our root and intermediary CA certs into the WSL machine to `\\wsl$\rancher-desktop\usr\local\share\ca-certificates`. From windows terminal I then ran `wsl -d rancher-desktop -e update-ca-certificates` and confirmed that my certs now appeared in `\\wsl$\rancher-desktop\etc\ss\certs\`. I then quit and started rancher desktop after a few second pause. Then I was able to pull the images again without the daemon.json file. It would be nice if somehow at startup it was able to get the certs our local machine already trusts but this is easy enough right now.

**Comment by jandubois** [comment]
I have not been able to reproduce this issue (with Rancher Desktop 1.0.1).

I've setup a private registry with a self-signed cert and basic auth requirements as `registry.home`. It contains an `nginx:test` image that I will use as a base image:

```console
C:\Users\Jan\sample>nerdctl build -t testing:latest .
[+] Building 0.2s (3/3) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.1s
 => => transferring dockerfile: 95B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.1s
 => => transferring context: 2B                                                                                    0.0s
 => ERROR [internal] load metadata for registry.home/nginx:test                                                    0.1s
------
 > [internal] load metadata for registry.home/nginx:test:
------
Dockerfile:1
--------------------
   1 | >>> FROM registry.home/nginx:test
   2 |     ADD Dockerfile DOCKERFILE
   3 |
--------------------
error: failed to solve: registry.home/nginx:test: failed to do request: Head "https://registry.home/v2/nginx/manifests/test": x509: certificate signed by unknown authority
FATA[0000] unrecognized image format
```

As expected, this failed because I have not yet installed the CA that I used to sign the certificate.

So I use the "Certificate Import Wizard" (from "Control Panel") to add my CA to the trust store:

<img width="477" alt="Screen Shot 2022-02-07 at 3 57 15 PM" src="https://user-images.githubusercontent.com/78947/152892325-de884abe-6024-4869-8ba1-abd77b91eb4e.png">

Afterwards I stop Rancher Desktop (via "File | Exit") and start it again, because certs are only imported into the distro during startup. Then I try again:

```console
C:\Users\Jan\sample>nerdctl build -t testing:latest .
[+] Building 0.3s (3/3) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.1s
 => => transferring dockerfile: 95B                                                                                0.1s
 => [internal] load .dockerignore                                                                                  0.1s
 => => transferring context: 2B                                                                                    0.0s
 => ERROR [internal] load metadata for registry.home/nginx:test                                                    0.1s
------
 > [internal] load metadata for registry.home/nginx:test:
------
Dockerfile:1
--------------------
   1 | >>> FROM registry.home/nginx:test
   2 |     ADD Dockerfile DOCKERFILE
   3 |
--------------------
error: failed to solve: registry.home/nginx:test: pulling from host registry.home failed with status code [manifests test]: 401 Unauthorized
FATA[0001] unrecognized image format
```

It still fails, because I haven't authenticated with the registry yet, but the certificate error is gone. So I log in, and try once more:

```console
C:\Users\Jan\sample>nerdctl login registry.home
Enter Username: user
Enter Password:
Login Succeeded

C:\Users\Jan\sample>nerdctl build -t testing:latest .
[+] Building 5.3s (8/8) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.1s
 => => transferring dockerfile: 95B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for registry.home/nginx:test                                                          0.1s
 => [auth] sharing credentials for registry.home                                                                   0.0s
 => [internal] load build context                                                                                  0.0s
 => => transferring context: 31B                                                                                   0.0s
 => [1/2] FROM registry.home/nginx:test@sha256:ee89b00528ff4f02f2405e4ee221743ebc3f8e8dd0bfd5c4c20a2fa2aaa7ede3    0.0s
 => => resolve registry.home/nginx:test@sha256:ee89b00528ff4f02f2405e4ee221743ebc3f8e8dd0bfd5c4c20a2fa2aaa7ede3    0.0s
 => CACHED [2/2] ADD Dockerfile DOCKERFILE                                                                         0.0s
 => exporting to oci image format                                                                                  5.1s
 => => exporting layers                                                                                            0.0s
 => => exporting manifest sha256:b2cc4ecafca7094099f83a5e08612c7ecc542040cccb993ef01243dd63406d40                  0.0s
 => => exporting config sha256:a231badb0e5071c5ce2bac5220ad538295ca8b4ae40aa983af0b1dc251fbb2a9                    0.0s
 => => sending tarball                                                                                             5.0s
unpacking docker.io/library/testing:latest (sha256:b2cc4ecafca7094099f83a5e08612c7ecc542040cccb993ef01243dd63406d40)...done

C:\Users\Jan\sample>nerdctl images
REPOSITORY    TAG       IMAGE ID        CREATED           PLATFORM       SIZE
testing       latest    b2cc4ecafca7    17 minutes ago    linux/amd64    149.1 MiB
```

Everything is as I expected.

If you still have problems connecting, can you follow the same steps and report back where you get different results?

**Comment by jandubois** [comment]
I know this Github issue is just about `nerdctl build`, but out of curiosity I switched the container runtime to `dockerd` and tried with the `docker` cli, and it also worked fine:

```console
C:\Users\Jan\sample>docker images
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE

C:\Users\Jan\sample>docker login registry.home
Authenticating with existing credentials...
WARNING! Your password will be stored unencrypted in C:\Users\Jan\.docker\config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Login Succeeded

C:\Users\Jan\sample>docker build -t testing:latest .
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM registry.home/nginx:test
test: Pulling from nginx
a2abf6c4d29d: Pull complete
a9edb18cadd1: Pull complete
589b7251471a: Pull complete
186b1aaa4aa6: Pull complete
b4df32aa5a72: Pull complete
a0bcbecc962e: Pull complete
Digest: sha256:ee89b00528ff4f02f2405e4ee221743ebc3f8e8dd0bfd5c4c20a2fa2aaa7ede3
Status: Downloaded newer image for registry.home/nginx:test
 ---> 605c77e624dd
Step 2/2 : ADD Dockerfile DOCKERFILE
 ---> 562525531cdc
Successfully built 562525531cdc
Successfully tagged testing:latest
SECURITY WARNING: You are building a Docker image from Windows against a non-Windows Docker host. All files and directories added to build context will have '-rwxr-xr-x' permissions. It is recommended to double check and reset permissions for sensitive files and directories.

C:\Users\Jan\sample>docker images
REPOSITORY                         TAG       IMAGE ID       CREATED         SIZE
testing                            latest    562525531cdc   9 seconds ago   141MB
registry.home/nginx                test      605c77e624dd   5 weeks ago     141MB
rancher/local-path-provisioner     v0.0.21   fb9b574e03c3   6 weeks ago     35MB
rancher/mirrored-metrics-server    v0.5.2    f73640fb5061   2 months ago    64.3MB
rancher/mirrored-coredns-coredns   1.8.6     a4ca41631cc7   4 months ago    46.8MB
rancher/mirrored-pause             3.6       6270bb605e12   5 months ago    683kB
```

**Comment by jandubois** [comment]
I assume this issue is resolved; please open a new one with additional information if you still have problems using a private registry with Rancher Desktop.

**Comment by Elektronenvolt** [comment]
@jandubois  I'm running in exactly this problem with RD 1.0.1 - started to describe it short and in a way you can reproduce it. I'll create a new issue.
